### PR TITLE
OP-320: Add possibility to select section type and fill up collection of items of this type

### PR DESCRIPTION
**Pull request by jkindly**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | OP-320, OP-312
| License         | MIT

### CODE REVIEW

File: spec/Assigner/CollectionsAssignerSpec.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
```
Comment (senghe) [inline spec/Assigner/CollectionsAssignerSpec.php:1]: Missing BitBag OpenSource header comment. Could you add this to every file on which it is missing?

File: spec/Assigner/CollectionsAssignerSpec.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public function it_implements_collections_assigner_interface(): void
+    {
+        $this->shouldHaveType(CollectionsAssignerInterface::class);
+    }
+
+    public function it_assigns_collections(
+        CollectionRepositoryInterface $collectionRepository,
+        CollectionInterface           $aboutCollection,
+        CollectionInterface           $blogCollection,
+        CollectionableInterface       $collectionsAware
```
Comment (senghe) [inline spec/Assigner/CollectionsAssignerSpec.php:35]: We don't use that style of indentation.

File: spec/Importer/BlockImporterSpec.php [hunk 3]
```diff
@@ -24,18 +24,18 @@
...
 {
     public function let(
-        ResourceResolverInterface $blockResourceResolver,
-        LocaleContextInterface $localeContext,
-        ImporterSectionsResolverInterface $importerSectionsResolver,
-        ImporterChannelsResolverInterface $importerChannelsResolver,
-        ImporterProductsResolverInterface $importerProductsResolver,
-        ValidatorInterface $validator,
-        BlockRepositoryInterface $blockRepository
+        ResourceResolverInterface            $blockResourceResolver,
```
Comment (senghe) [inline spec/Importer/BlockImporterSpec.php:27]: Please fix the indentations in all files :pray:

File: spec/Resolver/ImporterCollectionsResolverSpec.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    public function it_is_initializable()
+    {
+        $this->shouldHaveType(ImporterCollectionsResolver::class);
+    }
+
+    public function it_resolves_collections_for_collectionable_entity(
+        CollectionsAssignerInterface $collectionsAssigner,
+        CollectionableInterface      $collectionable
+    )
+    {
```
Comment (senghe) [inline spec/Resolver/ImporterCollectionsResolverSpec.php:35]:
```suggestion
    ) {
```

File: src/Entity/Block.php [hunk 5]
```diff
@@ -17,7 +17,7 @@
 class Block implements BlockInterface
 {
     use ToggleableTrait;
-    use SectionableTrait;
+    use CollectionableTrait;
```
Comment (senghe) [inline src/Entity/Block.php:20]: Collectible

File: src/Entity/Block.php [hunk 6]
```diff
@@ -28,7 +28,7 @@ class Block implements BlockInterface
     public function __construct()
     {
         $this->initializeTranslationsCollection();
-        $this->initializeSectionsCollection();
+        $this->initializeCollectionsCollection();
```
Comment (senghe) [inline src/Entity/Block.php:31]: :joy:

File: src/Entity/BlockableTrait.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusCmsPlugin\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+
+trait BlockableTrait
```
Comment (senghe) [inline src/Entity/BlockableTrait.php:16]: BlocksCollectionTrait

File: src/Entity/BlockableInterface.php [hunk 8]
```diff
@@ -12,18 +12,15 @@
 
 use Doctrine\Common\Collections\Collection;
 
-interface SectionableInterface
+interface BlockableInterface
```
Comment (senghe) [inline src/Entity/BlockableInterface.php:15]: BlocksCollectionInterface

File: src/Entity/CollectionableInterface.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * We are hiring developers from all over the world. Join us and start your new, exciting adventure and become part of us: https://bitbag.io/career
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusCmsPlugin\Entity;
+
+use Doctrine\Common\Collections\Collection;
+
+interface CollectionableInterface
```
Comment (senghe) [inline src/Entity/CollectionableInterface.php:15]: Collectible

File: src/Entity/CollectionableTrait.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusCmsPlugin\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+
+trait CollectionableTrait
```
Comment (senghe) [inline src/Entity/CollectionableTrait.php:16]: Collectible

File: src/Entity/MediableInterface.php [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+ * We are hiring developers from all over the world. Join us and start your new, exciting adventure and become part of us: https://bitbag.io/career
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusCmsPlugin\Entity;
+
+use Doctrine\Common\Collections\Collection;
+
+interface MediableInterface
```
Comment (senghe) [inline src/Entity/MediableInterface.php:15]: MediaCollectionInterface

File: src/Entity/MediableTrait.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusCmsPlugin\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+
+trait MediableTrait
```
Comment (senghe) [inline src/Entity/MediableTrait.php:16]: MediaCollectionTrait

File: src/Entity/PageableInterface.php [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ * We are hiring developers from all over the world. Join us and start your new, exciting adventure and become part of us: https://bitbag.io/career
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusCmsPlugin\Entity;
+
+use Doctrine\Common\Collections\Collection;
+
+interface PageableInterface
```
Comment (senghe) [inline src/Entity/PageableInterface.php:15]: PagesCollectionInterface

File: src/Entity/PageableTrait.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+*/
+
+declare(strict_types=1);
+
+namespace BitBag\SyliusCmsPlugin\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+
+trait PageableTrait
```
Comment (senghe) [inline src/Entity/PageableTrait.php:16]: PagesCollectionTrait

File: src/Resources/config/doctrine/Page.orm.xml [hunk 15]
```diff
@@ -25,7 +25,7 @@
 
         <field name="publishAt" column="publish_at" type="datetime_immutable" nullable="true"/>
 
-        <many-to-many field="sections" target-entity="BitBag\SyliusCmsPlugin\Entity\SectionInterface">
+        <many-to-many field="collections" target-entity="BitBag\SyliusCmsPlugin\Entity\CollectionInterface">
```
Comment (senghe) [inline src/Resources/config/doctrine/Page.orm.xml:28]: How doctrine would behave with changing the field name, when we have a table with data? Isn't it safer to use the "sections" column name in all mapping places? The field can stay then as "collections".

File: tests/Integration/Repository/CollectionRepositoryTest.php [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+declare(strict_types=1);
+
+namespace Tests\BitBag\SyliusCmsPlugin\Integration\Repository;
+
+use ApiTestCase\JsonApiTestCase;
+use BitBag\SyliusCmsPlugin\Entity\CollectionInterface;
+use BitBag\SyliusCmsPlugin\Repository\CollectionRepositoryInterface;
+use Doctrine\ORM\QueryBuilder;
+
+class CollectionRepositoryTest extends JsonApiTestCase
```
Comment (senghe) [inline tests/Integration/Repository/CollectionRepositoryTest.php:19]: Why not final?

**Review by senghe — CHANGES_REQUESTED** [review]

File: src/Resources/config/doctrine/Page.orm.xml (hunk 15 above)
Comment (jkindly) [reply src/Resources/config/doctrine/Page.orm.xml:28]: It doesn't change anything in the database. It is only related to the $collections field in entity, but in the database we have join-column-name: section_id

File: src/Entity/BlockableTrait.php (hunk 7 above)
Comment (senghe) [reply src/Entity/BlockableTrait.php:16]: This has not been changed yet

File: src/Entity/BlockableInterface.php (hunk 8 above)
Comment (senghe) [reply src/Entity/BlockableInterface.php:15]: huh?

File: src/Entity/CollectionableInterface.php (hunk 9 above)
Comment (senghe) [reply src/Entity/CollectionableInterface.php:15]: huh?

File: src/Entity/MediableInterface.php (hunk 11 above)
Comment (senghe) [reply src/Entity/MediableInterface.php:15]: huh?

**Review by senghe — CHANGES_REQUESTED** [review]
There are still some unresolved comments :vulcan_salute:

**Review by senghe — APPROVED** [review]

**Pull request merged**